XenModel/XenAPI-Extensions/VMSS.cs

```diff
@@ -0,0 +1,305 @@
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using XenAdmin;
+using XenAdmin.Alerts;
+using XenAdmin.Core;
+using XenAdmin.Network;
+using XenAdmin.Actions;
+
+namespace XenAPI
+{
+    public partial class VMSS : IVMPolicy
+    {
+        public bool is_enabled
+        {
+            get { return this.enabled; }
+        }
+
+        public bool is_running
+        {
+            get { return false; }
+        }
+
+        public bool is_archiving
+        {
+            get { return false; }
+        }
+        public DateTime _GetNextRunTime()
+        {
+            return this.GetNextRunTime();
+        }
+        public DateTime _GetNextArchiveRunTime()
+        {
+            return new DateTime();
+        }
+        public Type _Type
+        {
+            get { return typeof(VMSS); }
+        }
+        public List<PolicyAlert> PolicyAlerts
+        {
+            get { return Alerts; }
+        }
+
+        public bool hasArchive
+        {
+            get { return false; }
+        }
+
+        public void set_vm_policy(Session session, string _vm, string _value)
+        {
+            VM.set_snapshot_schedule(session, _vm, _value);
+        }
+
+        public void do_destroy(Session session, string _policy)
+        {
+            VMSS.destroy(session, _policy);
+        }
+
+        public string run_now(Session session, string _policy)
+        {
+            return VMSS.snapshot_now(session, _policy);
+        }
+
+        public void set_is_enabled(Session session, string _policy, bool _is_enabled)
+        {
+            VMSS.set_enabled(session, _policy, _is_enabled);
+        }
+
+        public PureAsyncAction getAlertsAction(IVMPolicy policy, int hoursfromnow)
+        {
+            return new GetVMSSAlertsAction((VMSS)policy, hoursfromnow);
+        }
+
+        public DateTime GetNextRunTime()
+        {
+            var time = Host.get_server_localtime(Connection.Session, Helpers.GetMaster(Connection).opaque_ref);
+
+            if (frequency == vmss_frequency.hourly)
+            {
+                return GetHourlyDate(time, Convert.ToInt32(backup_schedule_min));
+            }
+            if (frequency == vmss_frequency.daily)
+            {
+
+                var hour = Convert.ToInt32(backup_schedule_hour);
+                var min = Convert.ToInt32(backup_schedule_min);
+                return GetDailyDate(time, min, hour);
+
+            }
+            if (frequency == vmss_frequency.weekly)
+            {
+                var hour = Convert.ToInt32(backup_schedule_hour);
+                var min = Convert.ToInt32(backup_schedule_min);
+                return GetWeeklyDate(time, hour, min, new List<DayOfWeek>(DaysOfWeekBackup));
+            }
+            return new DateTime();
+
+        }
+
+        private static DateTime GetDailyDate(DateTime time, int min, int hour)
+        {
+            var nextDateTime = new DateTime(time.Year, time.Month, time.Day, hour, min, 0);
```

Comment:
In this file, any function which is static (or could be static like TryGetKey), we can just call the VMPP version directly, we don't need to make another copy of the code.